daemon/logs.go

```diff
@@ -14,13 +13,44 @@ import (
 	timetypes "github.com/docker/docker/api/types/time"
 	"github.com/docker/docker/container"
 	"github.com/docker/docker/daemon/logger"
-	"github.com/docker/docker/pkg/ioutils"
-	"github.com/docker/docker/pkg/stdcopy"
 )
 
-// ContainerLogs hooks up a container's stdout and stderr streams
-// configured with the given struct.
-func (daemon *Daemon) ContainerLogs(ctx context.Context, containerName string, config *backend.ContainerLogsConfig, started chan struct{}) error {
+// ContainerLogs copies the container's log channel to the channel provided in
+// the config. If ContainerLogs returns an error, no messages have been copied.
+// and the channel will be closed without data.
+//
+// if it returns nil, the config channel will be active and return log
+// messages until it runs out or the context is canceled.
+func (daemon *Daemon) ContainerLogs(ctx context.Context, containerName string, config *backend.ContainerLogsConfig) (rerr error) {
+
+	// O K so every path out of this function is an error before we start the
+	// streaming goroutine at the bottom, and nil after. if we return an error,
+	// that goroutine doesn't start (and doesn't close this channel, which, as
+	// the writer, it's our responsibility to close).
+	defer func() {
+		if rerr != nil {
+			close(config.Messages)
+		}
+	}()
+
+	// i know most of docker doesn't use logrus fields like this but swarmkit
+	// does and i think it's better so i might as well be the change i want to
+	// see in the world
+	lg := logrus.WithFields(logrus.Fields{
+		"module":    "daemon",
+		"method":    "(*Daemon).ContainerLogs",
+		"container": containerName,
+	})
+
+	// it's either error here or we will either block on trying to send, or
+	// panic when we try to close. the former case is gonna leave someone
+	// scratching their head for a few minutes before they realize their dumb
+	// mistake
+	if config.Messages == nil {
+		// i'm REALLY tempted to do panic() here because i NEVER get to use panic()
+		return errors.New("can't send messages on a nil channel, fix your code")
```

Comment:
Do you think it might be better to avoid this problem altogether by returning a channel, instead of relying on one being passed in?